packages/shared/src/rpc/index.ts

```diff
@@ -0,0 +1,181 @@
+import axios from 'axios';
+import { z } from 'zod';
+import { ChainflipNetwork, ChainflipNetworks } from '../enums';
+
+const numberOrHex = z
+  .union([z.string(), z.number()])
+  .transform((str) => BigInt(str));
+
+type CamelCase<T> = T extends string
+  ? T extends `${infer F}_${infer R}`
+    ? `${F}${Capitalize<CamelCase<R>>}`
+    : T
+  : never;
+
+type CamelCaseRecord<T> = T extends Record<string, unknown>
+  ? { [K in keyof T as CamelCase<K>]: CamelCaseRecord<T[K]> }
+  : T;
+
+const camelCase = <T extends string>(str: T): CamelCase<T> =>
+  str.replace(/_([a-z])/g, (_, char) => char.toUpperCase()) as CamelCase<T>;
+
+const camelCaseKeys = <T>(obj: T): CamelCaseRecord<T> => {
+  if (typeof obj !== 'object' || obj === null) return obj as CamelCaseRecord<T>;
+
+  return Object.fromEntries(
+    Object.entries(obj).map(([key, value]) => [
+      camelCase(key),
+      camelCaseKeys(value),
+    ]),
+  ) as CamelCaseRecord<T>;
+};
+
+const RPC_URLS: Record<ChainflipNetwork, string> = {
+  [ChainflipNetworks.backspin]: 'https://backspin-rpc.staging',
+  [ChainflipNetworks.sisyphos]: 'https://sisyphos.chainflip.xyz',
+  [ChainflipNetworks.perseverance]: 'https://perseverance.chainflip.xyz',
+  [ChainflipNetworks.mainnet]: 'https://mainnet.chainflip.io',
+};
+
+export type RpcConfig = { rpcUrl: string } | { network: ChainflipNetwork };
+
+const createRequest =
+  <P extends z.ZodTypeAny, R extends z.ZodTypeAny>(
+    method: string,
+    responseParser: R,
+    paramsParser?: P,
+  ) =>
+  async (
+    urlOrNetwork: RpcConfig,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    params: z.input<P> extends any ? void : z.input<P>,
+  ): Promise<CamelCaseRecord<z.output<R>>> => {
+    const url =
+      'network' in urlOrNetwork
+        ? RPC_URLS[urlOrNetwork.network]
+        : urlOrNetwork.rpcUrl;
+    const { data } = await axios.post(url, {
+      jsonrpc: '2.0',
+      method,
+      params: paramsParser?.parse(params),
+      id: 1,
+    });
+
+    const result = responseParser.safeParse(data.result);
+
+    if (result.success) {
+      return camelCaseKeys(result.data);
+    }
+
+    throw new Error(`RPC request "${method}" failed`, { cause: data.error });
+  };
+
+const fundingEnvironment = z.object({
+  redemption_tax: numberOrHex,
+  minimum_funding_amount: numberOrHex,
+});
+export const getFundingEnvironment = createRequest(
+  'cf_funding_environment',
+  fundingEnvironment,
+);
+
+const chainAssetMap = <Z extends z.ZodTypeAny>(parser: Z) =>
+  z.object({
```

Comment:
do we need to hardcode these object shapes or is there a way we could generate this (and the `rpcAsset`/`feesInfo`) from the constants we have?

Reply:
its okay for me to merge it like this

Reply:
I thought about trying to generate it but I don't think we will be adding assets fast enough to warrant the time it would take to try to get the zod parser to output exactly what we want